[UnifiedPDF] PDF content is offset in snapshots

smfr opened the pull request:
#### 31001e9814a15c1e2639a8fb4836b1ec65cc2fd1
<pre>
[UnifiedPDF] PDF content is offset in snapshots
<a href="https://bugs.webkit.org/show_bug.cgi?id=266791">https://bugs.webkit.org/show_bug.cgi?id=266791</a>
<a href="https://rdar.apple.com/120014898">rdar://120014898</a>

Reviewed by Tim Horton.

UnifiedPDFPlugin had two issues related to painting; first, it painted the PDF content as well
as rendering it into a layer. Second, that painted content was in the wrong location.

Fix the first issue by checking for a snapshotting paint in PluginView::paint(), and returning
if not snapshotting and the plugin uses PluginLayerHostingStrategy::GraphicsLayer, which is only
true for the UnifiedPDFPlugin. Also factor the PDF painting into a new function,
UnifiedPDFPlugin::paintPDFContent(), which avoids the confusion of passing in m_contentsLayer
when doing a software paint.

The second issue is also fixed in PluginView::paint(); we have to undo the context transform
that RenderWidget::paintContents() does.

Minor cleanup in PluginView::invalidateRect() to use a Checked&lt;&gt; pointer.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paint):
(WebKit::UnifiedPDFPlugin::paintContents):
(WebKit::UnifiedPDFPlugin::paintPDFContent):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::protectedPlugin const):
(WebKit::PluginView::paint):
(WebKit::PluginView::checkedRenderer const):
(WebKit::PluginView::invalidateRect):
* Source/WebKit/WebProcess/Plugins/PluginView.h:

Canonical link: <a href="https://commits.webkit.org/272449@main">https://commits.webkit.org/272449@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a19cad1864cdb4bb549636bde8fa787247fe9217

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/31624 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10303 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33339 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34115 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/28640 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/12657 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/7553 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/28241 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31972 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8676 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/28222 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7483 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7642 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35460 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28741 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/28579 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33781 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7736 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5744 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31633 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9392 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7427 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8421 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8244 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PluginView.cpp:
```diff
@@ -534,7 +532,28 @@ void PluginView::paint(GraphicsContext& context, const IntRect& dirtyRect, Widge
+        return frameView->paintBehavior().contains(PaintBehavior::FlattenCompositingLayers);
+    }();
+
+    if (!isSnapshotting && m_plugin->layerHostingStrategy() == PluginLayerHostingStrategy::GraphicsLayer)
```
> ... and the plugin uses PluginLayerHostingStrategy::GraphicsLayer, which is only
true for the UnifiedPDFPlugin

Would an `m_plugin->isUnifiedPDFPlugin()` be appropriate here? Or we want to do this early return whenever a plugin has that layer hosting strategy (not familiar with what this means, so this could be an unsound comment...)

smfr replied:
`isUnifiedPDFPlugin` would be also fine, and equivalent. But maybe some day we'll have a 3rd plugin type that uses GraphicsLayers.

aprotyas reviewed (commented):
LGTM. Small typo in the commit message:

> Second, that paainted content was in the wrong location.

"painted"

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -182,6 +182,7 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     float deviceScaleFactor() const override;
     float pageScaleFactor() const override;
 
+    void paintPDFContent(WebCore::GraphicsContext&, WebCore::FloatRect clipRect);
```
`const WebCore::FloatRect& clipRect`?

cdumez commented on Source/WebKit/WebProcess/Plugins/PluginView.cpp:
```diff
@@ -840,6 +859,11 @@ void PluginView::redeliverManualStream()
 
+CheckedPtr<RenderEmbeddedObject> PluginView::checkedRenderer() const
+{
+    return dynamicDowncast<RenderEmbeddedObject>(m_pluginElement->renderer());
```
Can the renderer be of any other type? If not, you could consider calling `checkedDowncast<>()` instead.

smfr replied:
`HTMLPlugInElement` uses `dynamicDowncast<RenderEmbeddedObject>` everywhere, so it's hard to know.

cdumez commented on Source/WebKit/WebProcess/Plugins/PluginView.cpp:
```diff
@@ -534,7 +532,28 @@ void PluginView::paint(GraphicsContext& context, const IntRect& dirtyRect, Widge
+    auto paintRect = dirtyRect;
+    paintRect.moveBy(-widgetOrigin);
+
+    m_plugin->paint(context, paintRect);
```
We should have a RefPtr to the plugin on the stack before we call paint() on it. something like:
```
protectedPlugin()->paint(...);
```

cdumez commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -306,6 +305,11 @@
+    paintPDFContent(context, clipRect);
+}
+
+void UnifiedPDFPlugin::paintPDFContent(WebCore::GraphicsContext& context, WebCore::FloatRect clipRect)
```
`const WebCore::FloatRect&` here too.

cdumez commented on Source/WebKit/WebProcess/Plugins/PluginView.cpp:
```diff
@@ -850,13 +874,12 @@ void PluginView::invalidateRect(const IntRect& dirtyRect)
 
-    auto* renderer = m_pluginElement->renderer();
-    if (!is<RenderEmbeddedObject>(renderer))
+    auto renderer = checkedRenderer();
```
We usually call `checkedX()` / `protectedX()` functions only to *avoid* local variables. When we have a local variable, we often use something like:
```
CheckedPtr renderer = m_renderer;
```

Not the explicit use of a smart pointer instead of auto. Sadly, checkedRenderer() also does a cast internally.
Maybe we should have a `renderer()` method which does the cast and use this here:
```
CheckedPtr renderer = this->renderer();
```

smfr replied:
But I want the `dynamicDowncast<RenderEmbeddedObject>()` that `checkedRenderer()` does

smfr committed:
[UnifiedPDF] PDF content is offset in snapshots
https://bugs.webkit.org/show_bug.cgi?id=266791
rdar://120014898

Reviewed by Tim Horton.

UnifiedPDFPlugin had two issues related to painting; first, it painted the PDF content as well
as rendering it into a layer. Second, that painted content was in the wrong location.

Fix the first issue by checking for a snapshotting paint in PluginView::paint(), and returning
if not snapshotting and the plugin uses PluginLayerHostingStrategy::GraphicsLayer, which is only
true for the UnifiedPDFPlugin. Also factor the PDF painting into a new function,
UnifiedPDFPlugin::paintPDFContent(), which avoids the confusion of passing in m_contentsLayer
when doing a software paint.

The second issue is also fixed in PluginView::paint(); we have to undo the context transform
that RenderWidget::paintContents() does.

Minor cleanup in PluginView::invalidateRect() to use a Checked<> pointer.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paint):
(WebKit::UnifiedPDFPlugin::paintContents):
(WebKit::UnifiedPDFPlugin::paintPDFContent):
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::protectedPlugin const):
(WebKit::PluginView::paint):
(WebKit::PluginView::checkedRenderer const):
(WebKit::PluginView::invalidateRect):
* Source/WebKit/WebProcess/Plugins/PluginView.h:

Canonical link: https://commits.webkit.org/272449@main
